### PR TITLE
Added LMDB support (plus some cleanup)

### PR DESCRIPTION
**Pull request by taktoa**
Still working on this. I recommend reviewing on a commit-by-commit basis (the whole patch is a bit messy to read).

### CODE REVIEW

File: src/Network/DFINITY/RadixTree.hs [hunk 1]
```diff
@@ -120,7 +120,7 @@ createRadixTree bloomSize cacheSize checkpoint database
 isEmptyRadixTree
    :: RadixTree database -- ^ Radix tree.
    -> Bool
-isEmptyRadixTree = (==) defaultRoot . _radixRoot
+isEmptyRadixTree db = _radixRoot db == defaultRoot
```
Comment (enzoh) [inline src/Network/DFINITY/RadixTree.hs]: I think we should try to stay consistent with naming variables. We're talking about the tree, not the database here.

Comment (taktoa) [reply src/Network/DFINITY/RadixTree.hs]: Good point.

File: src/Network/DFINITY/RadixTree/Types.hs [hunk 2]
```diff
@@ -70,9 +70,9 @@ type RadixBuffer = Map RadixRoot RadixNode
 
 type RadixCache = LruCache RadixRoot RadixNode
 
-class Monad m => RadixDatabase m database where
-   load :: database -> ByteString -> m (Maybe ByteString)
-   store :: database -> ByteString -> ByteString -> m ()
+class Monad m => RadixDatabase m db where
+   load :: db -> ByteString -> m (Maybe ByteString)
+   store :: db -> ByteString -> ByteString -> m ()
```
Comment (enzoh) [inline src/Network/DFINITY/RadixTree/Types.hs]: I suppose I have a habit of using verbose names. If we want to make cosmetic changes like this, then we should be consistent about it. This convention appears throughout this library, as well as other packages, including `dfinity-hypervisor`, `dfinity-p2p`, and `dfinity-client`.

Comment (taktoa) [reply src/Network/DFINITY/RadixTree/Types.hs]: I'm not wedded to `db`, and you make a fair point, so I'll revert this.

File: src/Network/DFINITY/RadixTree.hs [hunk 3]
```diff
@@ -606,44 +623,53 @@ contentsRadixTree' flag strategy = \ tree@RadixTree {..} -> do
                   let tree' = setRoot root tree
                   let prefix'' = prefix' ++ [test]
                   loop tree' prefix'' accum''
+{-# SPECIALISE contentsRadixTree'
```
Comment (enzoh) [inline src/Network/DFINITY/RadixTree.hs]: I suppose for consistency we would add a new line before the specialize block?

Comment (taktoa) [reply src/Network/DFINITY/RadixTree.hs]: Yeah, definitely; that was a typo.

Comment (enzoh) [reply src/Network/DFINITY/RadixTree.hs]: There are also some specialize blocks in `Network.DFINITY.RadixTree.Memory`. Maybe we want to modify those too. You're right, the standard practice is to put theses after the function.

Comment (enzoh) [reply src/Network/DFINITY/RadixTree.hs]: Could add specialization for LMDB too...

**Comment by taktoa** [comment]
@enzoh yes, otherwise the lmdb-simple dependency fails to build (since it has to link into LMDB)

**Pull request merged**